netket/experimental/observable/variance/expect.py

```diff
@@ -0,0 +1,137 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import jax.numpy as jnp
+import jax
+from functools import partial
+
+from netket.vqs import (
+    MCState,
+    expect,
+    expect_and_grad,
+    get_local_kernel,
+    get_local_kernel_arguments,
+)
+import netket.jax as nkjax
+from netket.utils import mpi
+
+from .variance_operator import VarianceOperator
+
+
+@expect.dispatch
+def expect(vstate: MCState, varop: VarianceOperator, chunk_size: None):
+    if varop.hilbert != vstate.hilbert:
+        raise TypeError("Hilbert spaces should match")
+
+    local_kernel = get_local_kernel(vstate, varop.op)
+    local_kernel2 = get_local_kernel(vstate, varop.op2)
+
+    sigma, args = get_local_kernel_arguments(vstate, varop.op)
+    sigma, args2 = get_local_kernel_arguments(vstate, varop.op2)
+
+    return expect_and_grad_inner_mc(
+        vstate._apply_fun,
+        vstate.parameters,
+        vstate.model_state,
+        local_kernel,
+        local_kernel2,
+        sigma,
+        args,
+        args2,
+        return_grad=False,
+    )
+
+
+@expect_and_grad.dispatch
+def expect_and_grad(
+    vstate: MCState,
+    varop: VarianceOperator,
+    chunk_size: None,
+    *,
+    mutable,
+):
+    if varop.hilbert != vstate.hilbert:
+        raise TypeError("Hilbert spaces should match")
+
+    local_kernel = get_local_kernel(vstate, varop.op)
+    local_kernel2 = get_local_kernel(vstate, varop.op2)
+
+    sigma, args = get_local_kernel_arguments(vstate, varop.op)
+    sigma, args2 = get_local_kernel_arguments(vstate, varop.op2)
+
+    return expect_and_grad_inner_mc(
+        vstate._apply_fun,
+        vstate.parameters,
+        vstate.model_state,
+        local_kernel,
+        local_kernel2,
+        sigma,
+        args,
+        args2,
+        return_grad=True,
+    )
+
+
+@partial(
+    jax.jit, static_argnames=("afun", "local_kernel", "local_kernel2", "return_grad")
+)
+def expect_and_grad_inner_mc(
+    afun,
+    params,
+    model_state,
+    local_kernel,
+    local_kernel2,
+    sigma,
+    args,
+    args2,
+    return_grad,
+):
+    n_chains = sigma.shape[0]
+
+    σ = sigma.reshape(-1, sigma.shape[-1])
+
+    def expect_kernel_var(params):
+        log_pdf = lambda params, σ: 2 * afun({"params": params, **model_state}, σ).real
+
+        def local_kernel_var(params, σ):
+            W = {"params": params, **model_state}
+
+            O_mean = nkjax.expect(
+                log_pdf,
+                lambda params, σ: local_kernel(
+                    afun, {"params": params, **model_state}, σ, args
+                ),
+                params,
+                σ,
+                n_chains=n_chains,
+            )[0]
+
+            O2_loc = local_kernel2(afun, W, σ, args2)
```

Comment:
we should take the real part here? the variance should be real

Reply:
True